shakenfist/daemons/cluster.py

```diff
@@ -37,37 +42,175 @@ class Monitor(daemon.Daemon):
 
+        # Inspect current state of blobs, the actual changes are done below outside
+        # the read only cache.
+        overreplicated = {}
+        underreplicated = []
+        nodes_by_disk_ascending = sorted(
+            nodes_by_free_disk_descending(intention='blobs'), reverse=True)
+        low_disk_nodes = nodes_by_free_disk_descending(
+            minimum=0, maximum=config.MINIMUM_FREE_DISK,
+            intention='blobs')
+
+        absent_nodes = []
+        for n in Nodes([node_inactive_states_filter]):
+            LOG.with_fields({
+                'node': n.fqdn}).info('Node is absent for blob replication')
+            absent_nodes.append(n.fqdn)
+        LOG.info('Found %d inactive nodes' % len(absent_nodes))
+
+        current_fetches = defaultdict(list)
+        for workname, workitem in etcd.get_outstanding_jobs():
+            # A workname looks like: /sf/queue/sf-3/jobname
+            _, _, phase, node, _ = workname.split('/')
+            if node == 'networknode':
+                continue
+
+            for task in workitem:
+                if isinstance(task, FetchBlobTask):
+                    if node in absent_nodes:
+                        LOG.with_fields({
+                            'blob': task.blob_uuid,
+                            'node': node,
+                            'phase': phase
+                        }).warning('Node is absent, ignoring fetch')
+                    else:
+                        LOG.with_fields({
+                            'blob': task.blob_uuid,
+                            'node': node,
+                            'phase': phase
+                        }).info('Node is fetching blob')
+                        current_fetches[task.blob_uuid].append(node)
+
+        with etcd.ThreadLocalReadOnlyCache():
+            for b in blob.Blobs([active_states_filter]):
+                # If there is current work for a blob, we ignore it until that
+                # work completes
+                if b.uuid in current_fetches:
+                    LOG.with_fields({
+                        'blob': task.blob_uuid
+                    }).info('Blob has current fetches, ignoring')
+                    continue
+
+                locations = b.locations
+                ignored_locations = []
+                for n in absent_nodes:
+                    if n in locations:
+                        locations.remove(n)
+                        ignored_locations.append(n)
+
+                if ignored_locations:
+                    LOG.with_fields({
+                        'blob': b,
+                        'ignored_locations': ignored_locations
+                    }).info('Ignored some blob locations as nodes are absent')
+
+                delta = len(locations) - config.BLOB_REPLICATION_FACTOR
+                if delta > 0:
+                    excess_locations = b.locations
+                    in_use_locations = []
+
+                    for instance_uuid in b.instances:
```

Comment:
This line iterates every disk of every instance, and is repeated for every blob. Would be faster to build a dict of instances using each blob. I realise that the data is cached in memory.

Maybe leave that optimisation till later?

Reply:
Yeah, I feel like we both agree this is a bit big in terms of patch size now. Let's merge this and I'll do the optimization and the "allow excess" thing in another change.